File: go-version/leetcode/LC2506/similar_pairs.go
package LC2506

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/2/22 下午9:40
 * @FilePath: similar_pairs
 * @Description: 实现相似字符串数对
 */

func similarPairs(words []string) int {
	ans := 0
	mp := make(map[int]int)
	for _, ss := range words {
		mask := 0
		for _, c := range ss {
			mask |= 1 << (c - 'a')
		}
		ans += mp[mask]
		mp[mask]++
	}
	return ans
}

File: python-version/LC2506/similar_pairs.py
class Solution:
    def similarPairs(self, words: list[str]) -> int:
        ans = 0
        map = {}
        for ss in words:
            mask = 0
            for c in ss:
                mask |= 1 << (ord(c) - ord('a'))
            cnt = map.get(mask, 0)
            ans += cnt
            map[mask] = cnt+1
        return ans